src/rdkafka_sasl_oauthbearer_oidc.c

```diff
@@ -0,0 +1,529 @@
+
+        token_url = rk->rk_conf.sasl.oauthbearer.token_endpoint_url;
+
+        herr = rd_http_post_expect_json(rk, token_url, headers, post_fields,
+                                        post_fields_size, timeout_s, retry,
+                                        retry_ms, &json);
+
+        if (unlikely(herr != NULL)) {
+                rd_kafka_log(rk, LOG_ERR, "OIDC",
+                             "Failed to retrieve OIDC "
+                             "token from \"%s\": %s (%d)",
+                             token_url, herr->errstr, herr->code);
+                rd_kafka_oauthbearer_set_token_failure(rk, herr->errstr);
+                rd_http_error_destroy(herr);
+                goto done;
+        }
+
+        if (!cJSON_HasObjectItem(json, "access_token")) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON response with "
+                    "\"access_token\" field");
+                goto done;
+        }
+
+        parsed_token = cJSON_GetObjectItem(json, "access_token");
+
+        jwt_token = cJSON_GetStringValue(parsed_token);
+        if (jwt_token == NULL) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON "
+                    "response as a value string");
+                goto done;
+        }
+
+        decoded_payloads = rd_kafka_jwt_b64_decode(jwt_token);
+        payloads         = cJSON_Parse(decoded_payloads);
+
+        if (!cJSON_HasObjectItem(payloads, "exp")) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON response with "
+                    "\"exp\" field");
+                goto done;
+        }
+        jwt_exp = cJSON_GetObjectItem(payloads, "exp");
+        exp     = cJSON_GetNumberValue(jwt_exp);
+        if (isnan(exp)) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON response with "
+                    "valid \"exp\" field");
+                goto done;
+        }
+
+        if (!cJSON_HasObjectItem(payloads, "sub")) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON response with "
+                    "\"sub\" field");
+                goto done;
+        }
+        jwt_sub = cJSON_GetObjectItem(payloads, "sub");
+        sub     = cJSON_GetStringValue(jwt_sub);
+        if (sub == NULL) {
+                rd_kafka_oauthbearer_set_token_failure(
+                    rk,
+                    "Expected JSON response with "
+                    "valid \"sub\" field");
+                goto done;
+        }
+
+        extension_str = rk->rk_conf.sasl.oauthbearer.extensions_str;
+        extensions =
+            rd_string_split(extension_str, ',', rd_true, &extension_cnt);
+
+        extension_key_value = rd_kafka_conf_kv_split(
```

Comment:
this memory is never freed

Reply:
Run the tests with valgrind to find memory leaks. E.g., `TESTS=0126 ./run-tests.sh valgrind` (on linux)